cmd/thanos/query.go

```diff
@@ -536,3 +539,69 @@ func firstDuplicate(ss []string) string {
 
 	return ""
 }
+
+// engineFactory creates from 1 to 3 promql.Engines depending on
+// dynamicLookbackDelta and eo.LookbackDelta and returns a function
+// that returns appropriate engine for given maxSourceResolutionMillis.
+//
+// TODO: it seems like a good idea to tweak Prometheus itself
+// instead of creating several Engines here.
+func engineFactory(
+	newEngine func(promql.EngineOpts) *promql.Engine,
+	eo promql.EngineOpts,
+	dynamicLookbackDelta bool,
+) func(int64) *promql.Engine {
+	deltas := []int64{downsample.ResLevel0}
+	if dynamicLookbackDelta {
+		deltas = []int64{downsample.ResLevel0, downsample.ResLevel1, downsample.ResLevel2}
+	}
+	var (
+		engines      = make([]*promql.Engine, len(deltas))
+		ld           = eo.LookbackDelta.Milliseconds()
+		totalEngines int
+	)
+	wrapReg := func() prometheus.Registerer {
+		wr := extprom.WrapRegistererWith(map[string]string{"engine": strconv.Itoa(totalEngines)}, eo.Reg)
+		totalEngines++
+		return wr
+	}
+	rawEngine := newEngine(promql.EngineOpts{
```

Comment:
Can we remove this and init all engines in exactly the same way (no matter if raw or not raw) (:

Reply:
It might simplify the code

Reply:
Thanks for the suggestion. Done.